python/bridgestan/model.py

```diff
@@ -0,0 +1,507 @@
+
+        self._name = self.stanlib.name
+        self._name.restype = ctypes.c_char_p
+        self._name.argtypes = [ctypes.c_void_p]
+
+        self._param_num = self.stanlib.param_num
+        self._param_num.restype = ctypes.c_int
+        self._param_num.argtypes = [ctypes.c_void_p, ctypes.c_int, ctypes.c_int]
+
+        self._param_unc_num = self.stanlib.param_unc_num
+        self._param_unc_num.restype = ctypes.c_int
+        self._param_unc_num.argtypes = [ctypes.c_void_p]
+
+        self._param_names = self.stanlib.param_names
+        self._param_names.restype = ctypes.c_char_p
+        self._param_names.argtypes = [
+            ctypes.c_void_p,
+            ctypes.c_int,
+            ctypes.c_int,
+        ]
+
+        self._param_unc_names = self.stanlib.param_unc_names
+        self._param_unc_names.restype = ctypes.c_char_p
+        self._param_unc_names.argtypes = [ctypes.c_void_p]
+
+        self._param_constrain = self.stanlib.param_constrain
+        self._param_constrain.restype = ctypes.c_int
+        self._param_constrain.argtypes = [
+            ctypes.c_void_p,
+            ctypes.c_int,
+            ctypes.c_int,
+            double_array,
+            double_array,
+        ]
+
+        self._param_unconstrain = self.stanlib.param_unconstrain
+        self._param_unconstrain.restype = ctypes.c_int
+        self._param_unconstrain.argtypes = [ctypes.c_void_p, double_array, double_array]
+
+        self._param_unconstrain_json = self.stanlib.param_unconstrain_json
+        self._param_unconstrain_json.restype = ctypes.c_int
+        self._param_unconstrain_json.argtypes = [
+            ctypes.c_void_p,
+            ctypes.c_char_p,
+            double_array,
+        ]
+
+        self._log_density = self.stanlib.log_density
+        self._log_density.restype = ctypes.c_int
+        self._log_density.argtypes = [
+            ctypes.c_void_p,
+            ctypes.c_int,
+            ctypes.c_int,
+            double_array,
+            ctypes.POINTER(ctypes.c_double),
+        ]
+
+        self._log_density_gradient = self.stanlib.log_density_gradient
+        self._log_density_gradient.restype = ctypes.c_int
+        self._log_density_gradient.argtypes = [
+            ctypes.c_void_p,
+            ctypes.c_int,
+            ctypes.c_int,
+            double_array,
+            ctypes.POINTER(ctypes.c_double),
+            double_array,
+        ]
+
+        self._log_density_hessian = self.stanlib.log_density_hessian
+        self._log_density_hessian.restype = ctypes.c_int
+        self._log_density_hessian.argtypes = [
+            ctypes.c_void_p,
+            ctypes.c_int,
+            ctypes.c_int,
+            double_array,
+            ctypes.POINTER(ctypes.c_double),
+            double_array,
+            double_array,
+        ]
+
+        self._destruct = self.stanlib.destruct
+        self._destruct.restype = ctypes.c_int
+        self._destruct.argtypes = [ctypes.c_void_p]
+
+    @classmethod
```

Comment:
[comment]
I also didn't know you could declare "static" class functions like this.  The name "classmethod" is terrible, though, as that refers specifically to a non-static class function!

Reply:
There is also @staticmethod, which is the same except it doesn't receive `cls` as an input. In this case I suppose the distinction is arbitrary